File: app/src/main/java/dev/ragnarok/fenrir/db/impl/KeysRamStorage.java
package dev.ragnarok.fenrir.db.impl;

import android.util.SparseArray;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

import dev.ragnarok.fenrir.crypt.AesKeyPair;
import dev.ragnarok.fenrir.db.interfaces.IKeysStorage;
import dev.ragnarok.fenrir.db.interfaces.IStorages;
import dev.ragnarok.fenrir.util.Objects;
import dev.ragnarok.fenrir.util.Optional;
import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Maybe;
import io.reactivex.rxjava3.core.Single;


class KeysRamStorage implements IKeysStorage {

    private final SparseArray<List<AesKeyPair>> mData = new SparseArray<>();

    private List<AesKeyPair> prepareKeysFor(int accountId) {
        List<AesKeyPair> list = mData.get(accountId);
        if (Objects.isNull(list)) {
            list = new CopyOnWriteArrayList<>();
            mData.put(accountId, list);
        }

        return list;
    }

    @Override
    public Completable saveKeyPair(@NonNull AesKeyPair pair) {
        return Completable.create(e -> {
            prepareKeysFor(pair.getAccountId()).add(pair);
            e.onComplete();
        });
    }

    @Override
    public Single<List<AesKeyPair>> getAll(int accountId) {
        return Single.create(e -> {
            List<AesKeyPair> list = mData.get(accountId);
            List<AesKeyPair> result = new ArrayList<>(Objects.isNull(list) ? 0 : 1);
            if (Objects.nonNull(list)) {
                result.addAll(list);
            }

            e.onSuccess(result);
        });
    }

    @Override
    public Single<List<AesKeyPair>> getKeys(int accountId, int peerId) {
        return Single.create(e -> {
            List<AesKeyPair> list = mData.get(accountId);
            List<AesKeyPair> result = new ArrayList<>(Objects.isNull(list) ? 0 : 1);
            if (Objects.nonNull(list)) {
                for (AesKeyPair pair : list) {
                    if (pair.getPeerId() == peerId) {
                        result.add(pair);
                    }
                }
            }

            e.onSuccess(result);
        });
    }

    @Override
    public Single<Optional<AesKeyPair>> findLastKeyPair(int accountId, int peerId) {
        return Single.create(e -> {
            List<AesKeyPair> list = mData.get(accountId);
            AesKeyPair result = null;
            if (Objects.nonNull(list)) {
                for (AesKeyPair pair : list) {
                    if (pair.getPeerId() == peerId) {
                        result = pair;
                    }
                }
            }

            e.onSuccess(Optional.wrap(result));
        });
    }

    @Override
    public Maybe<AesKeyPair> findKeyPairFor(int accountId, long sessionId) {
        return Maybe.create(e -> {
            List<AesKeyPair> pairs = mData.get(accountId);
            AesKeyPair result = null;
            if (Objects.nonNull(pairs)) {
                for (AesKeyPair pair : pairs) {
                    if (pair.getSessionId() == sessionId) {
                        result = pair;
                        break;
                    }
                }
            }

            if (Objects.nonNull(result)) {
                e.onSuccess(result);
            }

            e.onComplete();
        });
    }

    @Override
    public Completable deleteAll(int accountId) {
        return Completable.create(e -> {
            mData.remove(accountId);
            e.onComplete();
        });
    }

    @Override
    public IStorages getStores() {
        throw new UnsupportedOperationException();
    }
}

File: app/src/main/kotlin/dev/ragnarok/fenrir/db/impl/KeysRamStorage.kt
package dev.ragnarok.fenrir.db.impl

import android.util.SparseArray
import dev.ragnarok.fenrir.crypt.AesKeyPair
import dev.ragnarok.fenrir.db.interfaces.IKeysStorage
import dev.ragnarok.fenrir.db.interfaces.IStorages
import dev.ragnarok.fenrir.util.Optional
import dev.ragnarok.fenrir.util.Optional.Companion.wrap
import io.reactivex.rxjava3.core.*
import java.util.concurrent.CopyOnWriteArrayList

internal class KeysRamStorage : IKeysStorage {
    private val mData = SparseArray<MutableList<AesKeyPair>>()
    private fun prepareKeysFor(accountId: Int): MutableList<AesKeyPair> {
        var list = mData[accountId]
        if (list == null) {
            list = CopyOnWriteArrayList()
            mData.put(accountId, list)
        }
        return list
    }

    override fun saveKeyPair(pair: AesKeyPair): Completable {
        return Completable.create { e: CompletableEmitter ->
            prepareKeysFor(pair.accountId).add(pair)
            e.onComplete()
        }
    }

    override fun getAll(accountId: Int): Single<List<AesKeyPair>> {
        return Single.create { e: SingleEmitter<List<AesKeyPair>> ->
            val list: List<AesKeyPair>? = mData[accountId]
            val result: MutableList<AesKeyPair> = ArrayList(if (list == null) 0 else 1)
            if (list != null) {
                result.addAll(list)
            }
            e.onSuccess(result)
        }
    }

    override fun getKeys(accountId: Int, peerId: Int): Single<List<AesKeyPair>> {
        return Single.create { e: SingleEmitter<List<AesKeyPair>> ->
            val list: List<AesKeyPair>? = mData[accountId]
            val result: MutableList<AesKeyPair> = ArrayList(if (list == null) 0 else 1)
            if (list != null) {
                for (pair in list) {
                    if (pair.peerId == peerId) {
                        result.add(pair)
                    }
                }
            }
            e.onSuccess(result)
        }
    }

    override fun findLastKeyPair(accountId: Int, peerId: Int): Single<Optional<AesKeyPair>> {
        return Single.create { e: SingleEmitter<Optional<AesKeyPair>> ->
            val list: List<AesKeyPair>? = mData[accountId]
            var result: AesKeyPair? = null
            if (list != null) {
                for (pair in list) {
                    if (pair.peerId == peerId) {
                        result = pair
                    }
                }
            }
            e.onSuccess(wrap(result))
        }
    }

    override fun findKeyPairFor(accountId: Int, sessionId: Long): Maybe<AesKeyPair> {
        return Maybe.create { e: MaybeEmitter<AesKeyPair> ->
            val pairs: List<AesKeyPair>? = mData[accountId]
            var result: AesKeyPair? = null
            if (pairs != null) {
                for (pair in pairs) {
                    if (pair.sessionId == sessionId) {
                        result = pair
                        break
                    }
                }
            }
            if (result != null) {
                e.onSuccess(result)
            }
            e.onComplete()
        }
    }

    override fun deleteAll(accountId: Int): Completable {
        return Completable.create { e: CompletableEmitter ->
            mData.remove(accountId)
            e.onComplete()
        }
    }

    override val stores: IStorages
        get() = throw UnsupportedOperationException()
}